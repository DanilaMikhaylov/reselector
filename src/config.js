'use strict'

const cosmiconfig = require('cosmiconfig')

const { config } = (cosmiconfig('reselector', { sync: true }).load() || {})

module.exports = Object.assign({
  env: false,
  prefix: 'data-',
  syntaxes: [
    '@babel/plugin-syntax-async-generators',
    '@babel/plugin-syntax-class-properties',
    '@babel/plugin-syntax-decorators',
    '@babel/plugin-syntax-dynamic-import',
    '@babel/plugin-syntax-export-default-from',
    '@babel/plugin-syntax-export-namespace-from',
    '@babel/plugin-syntax-flow',
    '@babel/plugin-syntax-function-bind',
    '@babel/plugin-syntax-import-meta',
    '@babel/plugin-syntax-jsx',
    '@babel/plugin-syntax-nullish-coalescing-operator',
    '@babel/plugin-syntax-numeric-separator',
    '@babel/plugin-syntax-object-rest-spread',
    '@babel/plugin-syntax-optional-catch-binding',
    '@babel/plugin-syntax-optional-chaining',
    '@babel/plugin-syntax-pipeline-operator',
    '@babel/plugin-syntax-throw-expressions',
  ],
}, config)
