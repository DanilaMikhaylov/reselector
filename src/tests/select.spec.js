import React from 'react'
import { select, resolve } from '../'


describe('select', () => {
  const Text = ({ children }) => <p>{children}</p>
  const Button = ({ children }) => <button><Text>{children}</Text></button>

  it('should build css selector', () => {
    const selector = select`${Button} > ${Text}`

    expect(selector).toMatchSnapshot()
    expect(select.css`${Button} > ${Text}`).toEqual(selector)
  })

  it('should build xpath selector', () => {
    expect(select.xpath`//*${Button}/${Text}`).toMatchSnapshot()
  })

  it('should work with variables', () => {
    const index = 0

    expect(select`${Button} > ${Text}:nth-child(${index})`).toMatchSnapshot()
  })
})

describe('select modifiers', () => {
  const componentName = 'Button'
  const path = require.resolve('./App')

  const defaultSelector = resolve(path)[componentName]

  const defaultSelectCssMod = selector => `${selector}[href]`
  const defaultSelectXpathMod = selector => `${selector}[href]`

  const selectExpectedCssMod = selector => `${select.css`${selector}`}[href]`
  const selectExpectedXpathMod = selector => `${select.xpath`${selector}`}[href]`

  // eslint-disable-next-line no-restricted-syntax, no-shadow
  for (const {
    title,
    selectCssMod,
    selectXpathMod,
    expectedCss,
    expectedXpath,
  } of Object.values([
      {
        title: 'should modify a selector when modifier is defined',

        selectCssMod: defaultSelectCssMod,
        selectXpathMod: defaultSelectXpathMod,

        expectedCss: selectExpectedCssMod(defaultSelector),
        expectedXpath: selectExpectedXpathMod(defaultSelector),
      },
      {
        title: 'should not modify a selector when modifier is not defined',

        selectCssMod: undefined,
        selectXpathMod: undefined,

        expectedCss: select.css`${defaultSelector}`,
        expectedXpath: select.xpath`${defaultSelector}`,
      },
    ])) {
    jest.resetModules()

    // eslint-disable-next-line no-shadow
    require('../config').selectors = (map, resolve) => {
      const Component = resolve(path)[componentName]

      map.css(Component, selectCssMod)
      map.xpath(Component, selectXpathMod)
    }

    const { css, xpath } = require('../select')

    it(title, () => {
      const selector = resolve(path)[componentName]

      expect(css`${selector}`).toBe(expectedCss)
      expect(xpath`${selector}`).toBe(expectedXpath)
    })
  }
})
