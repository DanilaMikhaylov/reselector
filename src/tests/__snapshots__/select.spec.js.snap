// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`select should build css selector 1`] = `"[data-30d26c9a] > [data-ffe55b71]"`;

exports[`select should build xpath selector 1`] = `"//[@data-30d26c9a]/[@data-ffe55b71]"`;
